// vite@3.2.0-beta.1/dist/node/cli.js
import { performance } from 'node:perf_hooks';
import { cac } from 'cac';
import colors from 'picocolors';
import { u as createLogger, e as resolveConfig } from './chunks/dep-78db20e1.js';
import { VERSION } from './constants.js';
import 'node:fs';
import 'node:path';
import 'node:url';
import 'node:module';
import '@rollup/plugin-alias';
import 'esbuild';
import '@rollup/plugin-commonjs';
import 'connect';
import 'cors';
import 'chokidar';
import 'launch-editor-middleware';
import 'picomatch';
import 'node:os';
import 'node:crypto';
import 'node:util';
import 'node:dns';
import 'resolve';
import '@ampproject/remapping';
import 'debug';
import '@rollup/pluginutils';
import 'etag';
import 'convert-source-map';
import 'node:buffer';
import 'mrmime';
import 'magic-string';
import 'es-module-lexer';
import 'tsconfck';
import 'resolve.exports';
import 'mlly';
import 'acorn';
import 'fast-glob';
import 'postcss-load-config';
import 'strip-literal';
import '@jridgewell/trace-mapping';
import 'http-proxy';
import 'node:http';
import 'node:https';
import 'ws';
import 'micromatch';
import 'ufo';
import 'json5';
import '@rollup/plugin-dynamic-import-vars';
import 'strip-ansi';
import 'sirv';
import 'periscopic';
import 'estree-walker';
import 'node:readline';
import 'connect-history-api-fallback';
import 'node:child_process';
import 'open';
import 'cross-spawn';
import 'node:zlib';
import 'dotenv';
import 'dotenv-expand';
import 'okie';

const cli = cac('vite');
/**
 * removing global flags before passing as command specific sub-configs
 */
function cleanOptions(options) {
    const ret = { ...options };
    delete ret['--'];
    delete ret.c;
    delete ret.config;
    delete ret.base;
    delete ret.l;
    delete ret.logLevel;
    delete ret.clearScreen;
    delete ret.d;
    delete ret.debug;
    delete ret.f;
    delete ret.filter;
    delete ret.m;
    delete ret.mode;
    return ret;
}
cli
    .option('-c, --config <file>', `[string] use specified config file`)
    .option('--base <path>', `[string] public base path (default: /)`)
    .option('-l, --logLevel <level>', `[string] info | warn | error | silent`)
    .option('--clearScreen', `[boolean] allow/disable clear screen when logging`)
    .option('-d, --debug [feat]', `[string | boolean] show debug logs`)
    .option('-f, --filter <filter>', `[string] filter debug logs`)
    .option('-m, --mode <mode>', `[string] set env mode`);
// dev
cli
    .command('[root]', 'start dev server') // default command
    .alias('serve') // the command is called 'serve' in Vite's API
    .alias('dev') // alias to align with the script name
    .option('--host [host]', `[string] specify hostname`)
    .option('--port <port>', `[number] specify port`)
    .option('--https', `[boolean] use TLS + HTTP/2`)
    .option('--open [path]', `[boolean | string] open browser on startup`)
    .option('--cors', `[boolean] enable CORS`)
    .option('--strictPort', `[boolean] exit if specified port is already in use`)
    .option('--force', `[boolean] force the optimizer to ignore the cache and re-bundle`)
    .action(async (root, options) => {
    // output structure is preserved even after bundling so require()
    // is ok here
    const { createServer } = await import('./chunks/dep-78db20e1.js').then(function (n) { return n.A; });
    try {
        const server = await createServer({
            root,
            base: options.base,
            mode: options.mode,
            configFile: options.config,
            logLevel: options.logLevel,
            clearScreen: options.clearScreen,
            optimizeDeps: { force: options.force },
            server: cleanOptions(options)
        });
        if (!server.httpServer) {
            throw new Error('HTTP server not available');
        }
        await server.listen();
        const info = server.config.logger.info;
        // @ts-ignore
        const viteStartTime = global.__vite_start_time ?? false;
        const startupDurationString = viteStartTime
            ? colors.dim(`ready in ${colors.reset(colors.bold(Math.ceil(performance.now() - viteStartTime)))} ms`)
            : '';
        info(`\n  ${colors.green(`${colors.bold('VITE')} v${VERSION}`)}  ${startupDurationString}\n`, { clear: !server.config.logger.hasWarned });
        server.printUrls();
    }
    catch (e) {
        createLogger(options.logLevel).error(colors.red(`error when starting dev server:\n${e.stack}`), { error: e });
        process.exit(1);
    }
});
// build
cli
    .command('build [root]', 'build for production')
    .option('--target <target>', `[string] transpile target (default: 'modules')`)
    .option('--outDir <dir>', `[string] output directory (default: dist)`)
    .option('--assetsDir <dir>', `[string] directory under outDir to place assets in (default: assets)`)
    .option('--assetsInlineLimit <number>', `[number] static asset base64 inline threshold in bytes (default: 4096)`)
    .option('--ssr [entry]', `[string] build specified entry for server-side rendering`)
    .option('--sourcemap', `[boolean] output source maps for build (default: false)`)
    .option('--minify [minifier]', `[boolean | "terser" | "esbuild"] enable/disable minification, ` +
    `or specify minifier to use (default: esbuild)`)
    .option('--manifest [name]', `[boolean | string] emit build manifest json`)
    .option('--ssrManifest [name]', `[boolean | string] emit ssr manifest json`)
    .option('--force', `[boolean] force the optimizer to ignore the cache and re-bundle (experimental)`)
    .option('--emptyOutDir', `[boolean] force empty outDir when it's outside of root`)
    .option('-w, --watch', `[boolean] rebuilds when modules have changed on disk`)
    .action(async (root, options) => {
    const { build } = await import('./chunks/dep-78db20e1.js').then(function (n) { return n.z; });
    const buildOptions = cleanOptions(options);
    try {
        await build({
            root,
            base: options.base,
            mode: options.mode,
            configFile: options.config,
            logLevel: options.logLevel,
            clearScreen: options.clearScreen,
            optimizeDeps: { force: options.force },
            build: buildOptions
        });
    }
    catch (e) {
        createLogger(options.logLevel).error(colors.red(`error during build:\n${e.stack}`), { error: e });
        process.exit(1);
    }
});
// optimize
cli
    .command('optimize [root]', 'pre-bundle dependencies')
    .option('--force', `[boolean] force the optimizer to ignore the cache and re-bundle`)
    .action(async (root, options) => {
    const { optimizeDeps } = await import('./chunks/dep-78db20e1.js').then(function (n) { return n.y; });
    try {
        const config = await resolveConfig({
            root,
            base: options.base,
            configFile: options.config,
            logLevel: options.logLevel
        }, 'build', 'development');
        await optimizeDeps(config, options.force, true);
    }
    catch (e) {
        createLogger(options.logLevel).error(colors.red(`error when optimizing deps:\n${e.stack}`), { error: e });
        process.exit(1);
    }
});
cli
    .command('preview [root]', 'locally preview production build')
    .option('--host [host]', `[string] specify hostname`)
    .option('--port <port>', `[number] specify port`)
    .option('--strictPort', `[boolean] exit if specified port is already in use`)
    .option('--https', `[boolean] use TLS + HTTP/2`)
    .option('--open [path]', `[boolean | string] open browser on startup`)
    .action(async (root, options) => {
    const { preview } = await import('./chunks/dep-78db20e1.js').then(function (n) { return n.B; });
    try {
        const server = await preview({
            root,
            base: options.base,
            configFile: options.config,
            logLevel: options.logLevel,
            mode: options.mode,
            preview: {
                port: options.port,
                strictPort: options.strictPort,
                host: options.host,
                https: options.https,
                open: options.open
            }
        });
        server.printUrls();
    }
    catch (e) {
        createLogger(options.logLevel).error(colors.red(`error when starting preview server:\n${e.stack}`), { error: e });
        process.exit(1);
    }
});
cli.help();
cli.version(VERSION);
cli.parse();
